index.js

```diff
@@ -401,46 +401,64 @@ Cartero.prototype.processMains = function( callback ) {
-			replace : function( file, match, theUrl ) {
-				theUrl = theUrl.trim();
+		_this.writeIndividualAssetsToDisk( newPackage, assetTypesToWriteToDisk, function( err ) {
+			if( err ) return _this.emit( 'error', err );
 
-				// absolute urls stay the same.
-				if( theUrl.charAt( 0 ) === '/' ) return match;
-				if( theUrl.indexOf( 'data:' ) === 0 ) return match; // data url, don't mess with this
+			newPackage.addTransform( replaceStringTransform, {
+				find : /url\(\s*[\"\']?([^)\'\"]+)\s*[\"\']?\s*\)/g,
+				replace : function( file, match, theUrl ) {
+					theUrl = theUrl.trim();
+
+					// absolute urls stay the same.
+					if( theUrl.charAt( 0 ) === '/' ) return match;
+					if( theUrl.indexOf( 'data:' ) === 0 ) return match; // data url, don't mess with this
+
+					var assetUrl = path.resolve( path.dirname( file ), theUrl );
+					var newAssetName = _this.assetMap[ assetUrl ];
+
+					if (newAssetName) {
+						// replace the url with the new name
+						theUrl = path.join( path.dirname( theUrl ), newAssetName );
+					} else {
+						// this happen when we have packages that have assets references that are not specified
+						// in the image tag in package.json. It happens in modules like jqueryui
+						log.warn( file, 'Asset reference ' + theUrl + ' cannot be resolved' );
+					}
 
-				var cssFilePathRelativeToPackageDir = path.relative( newPackage.path, file );
-				var cssFileDirPathRelativeToPackageDir = path.dirname( '/' + cssFilePathRelativeToPackageDir );
-				if( cssFileDirPathRelativeToPackageDir !== '/' ) cssFileDirPathRelativeToPackageDir += '/';
+					var cssFilePathRelativeToPackageDir = path.relative( newPackage.path, file );
+					var cssFileDirPathRelativeToPackageDir = path.dirname( '/' + cssFilePathRelativeToPackageDir );
+					if( cssFileDirPathRelativeToPackageDir !== '/' ) cssFileDirPathRelativeToPackageDir += '/';
 
-				// urls in css files are relative to the css file itself
-				var absUrl = url.resolve( cssFileDirPathRelativeToPackageDir, theUrl );
-				absUrl = _this.outputDirUrl + newPackage.id + absUrl;
+					// urls in css files are relative to the css file itself
+					var absUrl = url.resolve( cssFileDirPathRelativeToPackageDir, theUrl );
 
-				return 'url( \'' + absUrl + '\' )';
-			}
-		}, 'style' );
+					absUrl = _this.outputDirUrl + newPackage.id + absUrl;
 
-		newPackage.addTransform( assetUrlTransform, {
-			packagePathsToIds : _this.packagePathsToIds,
-			outputDirUrl : _this.outputDirUrl
-		}, 'style' );
+					return 'url( \'' + absUrl + '\' )';
+				}
+			}, 'style' );
 
-		_this.writeIndividualAssetsToDisk( newPackage, assetTypesToWriteToDisk, function( err ) {
-			if( err ) return _this.emit( 'error', err );
+			newPackage.addTransform( assetUrlTransform, {
+				packagePathsToIds : _this.packagePathsToIds,
+				outputDirUrl : _this.outputDirUrl,
+				assetMap: _this.assetMap
+			}, 'style' );
 
-			_this.emit( 'packageCreated', newPackage );
+			done();
```

Comment:
here is the magic, we can take all the time that we want to calculate the filenames and only then parcelify will continue processing